### ISSUE TITLE
XXX has or is using name 'Foo' from external module "../bar" but cannot be named

### ISSUE THREAD

**Issue by vilic** [Microsoft/TypeScript#9944]
**TypeScript Version:**  2.0.0

``` sh
npm init
npm install clime --save
```

**tsconfig.json**

``` json
{
    "compilerOptions": {
        "target": "es6",
        "module": "commonjs",
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "noEmitOnError": true,
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "strictNullChecks": true,
        "declaration": true,
        "sourceMap": true,
        "rootDir": "src",
        "outDir": "bld"
    },
    "exclude": [
        "node_modules"
    ]
}
```

**src/cli.ts**

``` ts
#!/usr/bin/env node

import * as Path from 'path';
import { CLI, Shim } from 'clime';

let cli = new CLI('henge', Path.join(__dirname, 'commands'));

// This is okay.
export default cli;

// This gives errors:
// TS4023: Exported variable 'foo' has or is using name 'Promise' from external module "C:/Projects/clime/node_modules/thenfail/bld/promise" but cannot be named.
// TS4023: Exported variable 'foo' has or is using name 'Printable' from external module "C:/Projects/clime/bld/core/object" but cannot be named.
let foo = cli.execute([]);
export { foo };
```

It seems that if I import `Printable` from `clime`, the second error would go away, but I cannot import `Promise` from the dependency of `clime`.

Is this expected behavior or a bug?

**Comment by mijay** [comment]
This (or very similar) issue actually completely breaks `declaration` feature in projects using external typings:

#### test.ts
```typescript
import { DOM, Component } from "react";
export class TestMe extends Component<{}, {}> {
// error TS4055: Return type of public method from exported class has or is using private name 'React.DOMElement'.
// error TS4055: Return type of public method from exported class has or is using private name 'React.HTMLAttributes'.
    render() {
        return DOM.div();
    }
}
```

#### tsconfig.json
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "moduleResolution": "node",
    "declaration": true
  },
  "files": [ "test.ts" ]
}
```

#### package.json
```json
{
  "name": "test",
  "version": "1.0.0",
  "dependencies": {
    "@types/react": "^0.14.31",
    "react": "^15.1.0"
  },
  "devDependencies": {
    "typescript": "^2.0.2"
  }
}
```

I can fix the errors by explicitly importing `DOMElement` and `HTMLAttributes`, but that negate all benefits of type inference - I am basically to force writing all types everywhere (especially with `noUnusedLocals`).

**Comment by mhegazy** [comment]
There is no explicit type annotation on the functions or variables. the declaration emitter infers their type and tries to write it. if the type is coming from a different module, then a. it needs to add an import or b. error.

The emitter can write the additional import, but that would have been changing your API shape in a way you did not indicate clearly in your code. so we opted to error instead.

the fix would be to add an explicit type annotation on the source of the problem.

Having siad that, i think we should reconsider this design decision, and add imports anyways.

**Comment by frederickfogerty** [comment]
Please consider adding adding imports to resolve this error. This is an error which comes up daily in our workflow with TypeScript and would be easier for us if this was never an issue again.

🙌

**Comment by landonpoch** [comment]
I was able to get around this issue by doing what @mhegazy suggested.  After adding an explicit type annotation (my return type was blank and having to get inferred from the parent class) then the errors went away.  Worked like a charm.  Thanks!

**Comment by ecraig12345** [comment]
My team at Microsoft is having this issue too, but in our case (a mapping of enum values to functions) it's not reasonable to add explicit types. Here's a greatly simplified version of our scenario:

tsconfig.json:
```
{
    "compilerOptions": {
        "module": "amd",
        "target": "es5",
        "declaration": true
    }
}
```

Mapping.ts:
```
import { fooFunction } from './FooFunction';
import { barFunction } from './BarFunction';

export enum Enum {
    foo,
    bar
}

export const mapping = {
    [Enum.foo]: fooFunction, // error TS4023: Exported variable 'mapping' has or is using name 'IFoo' from external module "D:/test/IFoo" but cannot be named.
    [Enum.bar]: barFunction // error TS4023: Exported variable 'mapping' has or is using name 'IBar' from external module "D:/test/IBar" but cannot be named.
}
```

FooFunction.ts:
```
export interface IFoo {
    bar: string;
}
export function fooFunction(foo: IFoo) {
    return foo.bar;
}
```

BarFunction.ts:
```
export interface IBar {
    foo: string;
}
export function barFunction(bar: IBar) {
    return bar.foo;
}
```

An manually written interface for the mapping would be incredibly ugly and hard to maintain (the actual mapping has 20+ values):
```
interface IMapping {
    "0": (foo: IFoo) => string;
    "1": (bar: IBar) => string;
}
```

**Comment by unional** [comment]
As a reminder that auto re-export or other mechanism is needed to properly resolve this issue. Adding import only is not sufficient.

https://github.com/Microsoft/TypeScript/issues/5711#issuecomment-231501080

### ----------- Original comment ----------

This could have deeper consequences.

Consider `moduleA` -> `moduleB` -> `moduleC` -> `moduleD`.

`moduleB` is the one that needs to do `import { ABC } from 'moduleA'` to get around this issue.

When `moduleC` uses `moduleB` and export its signature, it again fails to compile because `moduleB` needs `ABC` from `moduleA` but this time `moduleB` didn't export it.

This means either:
1. `moduleC` needs to have a hard dependency of `moduleA` and import `ABC`
2. `moduleB` needs to not just import but also re-export `ABC` and then `moduleC` imports it.

If `moduleD` does similar things, then basically you need to know the whole chain of dependencies.

~~I wasn't able to test this to prove that it is the case because I'm using `typings` and there is an issue blocking me so: https://github.com/typings/typings/issues/625~~

EDIT: I am able to reproduce it and indeed my `moduleD` needs to reference `moduleA` to import `ABC`.
In my example:

- `moduleA` => `redux`
- `moduleB` => `redux-thunk`
- `moduleC` => custom code on top of `redux-thunk`
- `moduleD` => some library
- `ABC` => `Dispatch` interface in `redux`

**Comment by mhegazy** [comment]
> An manually written interface for the mapping would be incredibly ugly and hard to maintain (the actual mapping has 20+ values):

you do not need an interface, you just need to export the imported types:


##### Mapping.ts:
```ts
....

export { fooFunction , barFunction };
```

**Comment by mischkl** [comment]
Just came across this issue today after upgrading Angular-CLI (uses AOT in production build, which apparently uses declarations flag).

Needless to say, from my perspective as a "naive" developer, the requirement that implicitly used interfaces be imported everywhere has the serious downside of poor developer tooling support. At least in WebStorm using the TypeScript service, the imported interfaces are shown as being unused and thus subject to removal by other developers or automatic removal whenever Optimize Imports is executed. Aside from that it is extremely non-obvious, from the wording of the error message, how the problem can be fixed.

**Comment by andy-ms** [comment]
Ran into this too. Here's a repro.
```ts
// @declaration: true
// @noUnusedLocals: true

// @Filename: a.ts
export class A {}
export var a: A;

// @Filename: b.ts
import { a, A } from "./a";
export const b = a;

// @Filename: c.ts
import { a } from "./a";
export const c = a;
```

This gets errors in both `b.ts` and `c.ts`.

**Comment by borislemke** [comment]
I ran into this issue when started to destructuring my `express` app into smaller modules.
Could someone kindly help me out how to cope with the issue in this example?

```typescript
import * as express from 'express'
...

export const WebsitePagesRouter = express.Router();

Error:
Exported variable "WebsitePagesRouter" has or is using name 'Router' from
external module '~path/to/@types/express-serve-static-core/index' but cannot be named.
```

**Comment by andy-ms** [comment]
@borislemke This error only happens when the type of an export is implicit, so try adding an explicit type declaration to `WebsitePagesRouter`.

**Comment by borislemke** [comment]
@andy-ms ahh now I got it. Thanks for the quick reply.
Simply adding the type fixed it
```
import {Router} from 'express'
export const MediaBundleRouter: Router = express.Router();
```

**Comment by gund** [comment]
@andy-ms you have a valid point but the real power of TS is that it is smart enough to resolve types implicitly based on return type of the function for example. And if you have a large codebase built on top of another one it is pointless to repeat types if they can be resolved implicitly.

I would suggest that it might be a good idea to make a feature like auto re-import in declaration file as an opt-in option in TS config file, something like `"implicitDeclaration": true`.

In this case default behavior will remain the same unless user nows what he wants and how it might modify his/her declaration files.

**Comment by captainjono** [comment]
Im new to typescript, but when a compiler complains to me that it cant find an import, because one dependency explicitly import its, but another imports it under an alias, thats just a bug, not ambiguity, or a feature, its a bug. And i can repo the above using aliases.

**Comment by mdekrey** [comment]
This is unfortunately a much larger issue with the new `noUnusedLocals` compiler switch; we can no longer just import the typing from the external module due to causing the new error `XXX is declared but never used`. Instead, we have to make sure the imported value is actually used somewhere.

**Comment by bfricka** [comment]
This is a particularly painful requirement limitation for lib code, and given the fact that ES2015 modules are statically defined, it's sad to see that a game plan (at least) for ameliorating this issue has yet to be defined.

**Comment by unional** [comment]
@mdekrey  an even harder case with this and `noUsusedLocals`:

```ts
// create.ts
export interface Foo { }

export function create(foo: Foo) { return foo }

// x.ts
// import { create } from './create' // missing private type Foo
import { create, Foo } from './create' // unused local Foo

export const abc = { create } // no place to use Foo !!
```

**Comment by paranoidjk** [comment]
+1 for this problem.

I do not think add an explicit type annotation is a good way, since it's hard to maintain.

**Comment by pauldraper** [comment]
Brainstorming: if the type hasn't been imported, tsc could output the structural type like `{a:number}` instead of `A`.

Would this make the issues any less?

**Comment by pauldraper** [comment]
> Having siad that, i think we should reconsider this design decision, and add imports anyways.

> A better and simpler fix can be done on the declaration emitter side, ensuring that it does inject these imports itself.

The problem I see is there are many possible paths for import. It could be absolute (`/home/paul/path/example.ts`, relative `../../../../example.ts`, non-relative with path mapping `path/example.ts`, or rely on node_modules `example.ts`).

TS has a long and detailed resolution order; it seems there would likewise have to be a reverse resolution order?

**Comment by atrauzzi** [comment]
What's the status of this? I can't even begin to list the number of times I've had to help people with this issue. At least several times a week in our projects.  There's something urgently wrong here.

**Comment by ryansmith94** [comment]
Agreed @atrauzzi. For me this has became the most annoying issue with TypeScript as of 2.4. What are the chances of there being a Compiler-based solution to this before 2.5 @mhegazy?

**Comment by ryansmith94** [comment]
This issue is particularly annoying when you want to use a union type from a different file, because you have to actually import the types that are part of the union everywhere you use the type.

**Comment by frederickfogerty** [comment]
To every commenting that they have the same issue, please 👍  the top comment. This allows the maintainers to easily see the magnitude of this issue. 

I know other Microsoft teams like VS Code use 👍 s to rank features/issues to be implemented/fixed.

**Comment by paprikka** [comment]
Not sure if it's the same issue, but I'm getting `TS4023` when using default exports with fat arrow functions, e.g.:
 

    // Module1.ts

    export interface ISomething {...}


    // Callback.ts

    // This won't work:
    import { ISomething } from 'Module1.ts'

    export default ( arg1: ISomething ): Promise<ISomething> => {
        ...
    }

    // This won't work:
    import { ISomething } from 'Module1.ts'

    const callback = ( arg1: ISomething ): Promise<ISomething> => {
        ...
    }
    export default callback




    // This will work:
    import { ISomething } from 'Module1.ts'

    export default function callback( arg1: ISomething ): Promise<ISomething> {
        ...
    }

**Comment by pkreiner** [comment]
An interesting quirk / tip for this is that, for exported functions involving types that cannot be named, sometimes 
`export const f = (x: X): Y = { ... };` 
will give `cannot be named` errors on types involved in `X` and `Y`, but 
`export const f: (x: X) => Y = (x) => { ... };`
will be fine. In other words, annotating with the argument and return types doesn't cause them to be re-exported properly, but annotating with the full function type does.

**Comment by stgolem** [comment]
Also vote for some kind of tsconfig flag for this. 
My team has to write many separated packages, that is based on some "core" packages.
This issue is like the hell of typescript.

**Comment by unional** [comment]
I just encounter another case that may be an edge case of this:

Given:

```ts
// src/column/column.ts
export interface Column { ... }

// src/column/index.ts
export * from './column'    
```

This code does not have error:

```ts
// src/grid/grid.ts
import { Column } from '../column'

export interface Grid {
  columns: Column[]
}

export function createGrid(): Grid { ... }
```

But this does:

```ts
// src/grid/interfaces.ts
import { Column } from '../column'

export interface Grid {
  columns: Column[]
}

// src/grid/createGrid.ts
import { Grid } from './interfaces'

export function createGrid(): Grid { ... } // error
```

The error would say something like:

> Return type of exported function has or is using name 'Column' from external module "c:/.../src/grid/interfaces" but cannot be named.

It is very likely a modified version of what I described: https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-254696587

And in this case, re-exporting doesn't make sense, because the whole library look and use like this:

```ts
// src/grid/index.ts
export * from './interfaces'
export * from './createGrid'

// src/index.ts
import * as column from './column'
import * as grid from './grid'

export { column, grid }

// usage
import { column, grid } from 'some-package'

let c = column.createColumn()
let g = grid.createGrid()

// usage in global namespace
import * from 'some-package'
export as namespace Abc

let c = Abc.column.createColumn()
let g = Abc.grid.createGrid()
```

It doesn't make sense to re-export `Column` under the `grid` namespace.

**Comment by mdekrey** [comment]
I started looking into this, but I'm afraid that my knowledge of the TypeScript codebase will hinder me from making a PR for this all on my own: it'd be my first PR to TypeScript, and this is still flagged as a "Suggestion" by @mhegazy anyway.

Reason for the change:

* Declaration emitting causes additional unclear errors for developers that can be automatically avoided.
* Declaration emitting with `noUnusedLocals` set to true can result in edge cases that are difficult to solve, especially while keeping the implicit nature of TypeScript clean.

My understanding of the proposal:

* A new flag, `shouldExportImplicitTypes`, should be added to `compilerOptions` to allow developers to opt-in to this case. (I personally would propose that this eventually is made the default for `"declaration": true`.)
* New import statements should be emitted in the declaration file _only_ to accomodate the implicit dependencies due to the types.
* No change should occur on the emitted JS files.

Changes that I believe need to happen:

* [`isSymbolAccessible` in checker.ts (approx line 2286 of the linked version)](https://github.com/Microsoft/TypeScript/blob/fd89808/src/compiler/checker.ts) will need to be updated to provide a new `SymbolAccessibilityResult` result that specifies the new import statement, possibly using `aliasesToMakeVisible`.
* A new `additionalTypeImports` map [similar to `usedTypeDirectiveReferences` would be added to declarationEmitter.ts](https://github.com/Microsoft/TypeScript/blob/fd8980830ca35f07d838de5457504b95e6974d3c/src/compiler/declarationEmitter.ts#L80) to allow injection of imports at the top of the file. Naming collisions would need to be observed and avoided, however these types would not be referenced anywhere outside of this file, since they weren't imported before and wouldn't be exported, so the naming could be something clearly auto-generated. Paths to the imported file would need to be relative and follow the module pathing conventions.
* Compiler flag would need to be added somehow; I didn't get that far.

**Comment by unional** [comment]
https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-254696587

Want to stress again that just importing does not solve the problem.
It's just kicking the can down a dead end (because the type is not re-exported, the consuming code have no way to import it without knowing the whole dependency chain).

**Comment by mdekrey** [comment]
I'm not sure I follow, @unional. The top of the output .d.ts will have the reference to the original export; TypeScript is tracking the location of the symbols by their original path already. 

From your original example:

> Consider `moduleA` -> `moduleB` -> `moduleC` -> `moduleD`.
>
> `moduleB` is the one that needs to do `import { ABC } from 'moduleA'` to get around this issue.
>
> When `moduleC` uses `moduleB` and export its signature, it again fails to compile because `moduleB` needs `ABC` from `moduleA` but this time `moduleB` didn't export it.

I believe your option 1 is the better solution, as it does not require re-exporting:
> `moduleC` needs to have a hard dependency of `moduleA` and import `ABC`

Re-exporting changes the contract of the module, including the JS in the case of types that also have emitted JS (such as classes). Adding an explicit dependency where there was previously an implicit dependency is a less breaking change (I could see an argument where it's a non-breaking change, but that's I believe the proposal should be opt-in), and only affects the typing... which is the source of the problem.

**Comment by jamiewinder** [comment]
The workaround for this at the moment seems to be to import the used types. However, unless you use this elsewhere in your code file (and have `--noUnusedLocals` enabled) you'll then be warned that you're importing something but aren't using it. 

Because of these two conflicting issues I have to awkwardly include some reference to the type in my files. For example, all of my styled-components code files that export their components have this abomination somewhere in the file just to keep TypeScript happy:

```typescript
((_?: StyledComponentClass<any, any>) => 0)();
```
I don't know if this is a genuine fix or just masking an underlying issue I don't understand.

The alternative being to have something like:
```typescript
const Header: StyledComponentClass<React.DetailedHTMLProps<React.HTMLAttributes<HTMLHeadingElement>, HTMLHeadingElement>, any, React.DetailedHTMLProps<React.HTMLAttributes<HTMLHeadingElement>, HTMLHeadingElement>>
 = styled.h1`
```
Which is even worse.

**Comment by tvald** [comment]
In order to work around this issue while checking for no unused locals, we use `tslint`'s equivalent [no-unused-variable](https://palantir.github.io/tslint/rules/no-unused-variable/) with an ignore prefix of underscore. The resulting type declaration import looks like:
```
import { SomeType as _S } from 'other-package'
```

Unlike `--noUnusedParameters`, `--noUnusedLocals` doesn't appear to support ignoring underscore-prefixed variable names. Adding that feature would enable the above workaround without falling back to `tslint`.

(This is still a painful workaround which in no way makes up for the loss of type inference, but at least we aren't blocked.)

**Comment by RyanCavanaugh** [comment]
@weswigham do you believe the new declaration emitter will fix this, or should we consider this as a standalone bug?

**Comment by weswigham** [comment]
It should be capable of it, but it's a new feature and right now I've gotten it focused on having parity with the current declaration emitter to make the change easier to initially merge.

**Comment by unional** [comment]
> I believe your option 1 is the better solution, as it does not require re-exporting:

It has some big consequences. One is about version mismatch.
Consider the following:

moduleB@0.3 depends on moduleA@0.1
moduleC@0.4 depends on moduleA@0.2
moduleD depends on moduleB@0.3 and moduleC@0.4

Now when moduleD wants to pass an interface of moduleA from moduleB code to moduleC code, you get: `Type 'X' is not assignable to type 'X'. Two different types with this name exist, but they are unrelated.`

**Comment by JakeGinnivan** [comment]
Any movement on this now the new emitter has been merged? This would allow us to cleanup so many useless imports/re-exports.

**Comment by simonbuchan** [comment]
@mhegazy a year ago you commented:
> There is no explicit type annotation on the functions or variables. the declaration emitter infers their type and tries to write it. if the type is coming from a different module, then a. it needs to add an import or b. error.
> 
> The emitter can write the additional import, but that would have been changing your API shape in a way you did not indicate clearly in your code. so we opted to error instead.

I'm not sure what "changing your API shape" here means, the imports of declaration files are opaque to importers, aren't they? The only exception I could think of is if A imports B, and B decides to add an import for C which has ambient declarations, then A would get unexpected, invisible in the source ambients?

The common case is something like:

```ts
// A.d.ts
export interface T { p: number, u: U };
export interface U { q: string };
export function f(): T;
```

```ts
// B.ts
import { f } from './A';
export const x = f();
```

Obviously here it's perfectly safe to emit:

```ts
// B.d.ts
import { T } from './A';
export declare const x: T;
```

If we then have:

```ts
// C.ts
import { x } from './B';
export const p = x.p;
```

then currently we get:

```ts
// C.d.ts
// hey, what? where's './B'?
export const p: number;
```

but:

```ts
// D.ts
import { x } from './B';
export const u = x.u; // hey, you can't use U here!
```

it seems you are saying that it would potentially have side effects to emit the intuitive:

```ts
// D.d.ts
import { U } from './A'; // hey, where did this come from?
export declare const u: U;
```

However, I can't see how a client of 'D' would be able to tell, since, of course, 'D.ts' already transitively imports 'A'. In fact, it seems like the current behavior of removing unused imports might cause ambient declarations in dependencies to "randomly" appear and disappear.

Since TS obviously already knows where the type that needs to be added is declared, the only issue I can think of left is *what* import should be added: how do you deal with imports containing `..`, `node_modules`, symlinks etc.? AFAICT TS currently deals with all of these by getting the absolute path, but you definitely don't want it emitting decl files by just taking relative paths, and even treating `node_modules` specially could get tripped up by weird stuff like `import publicVersion from '../../../node_modules/some-module'`.

As a side note, I think it's amusing that TS is perfectly happy with `export const u: typeof x.u = x.u;`, and emits:

```ts
import { x } from './B';
export declare const u: typeof x.u;
```

Of course, that's not really a path forward, since it would often end up with `typeof <500 line function invocation>`

**Comment by mhegazy** [comment]
> I'm not sure what "changing your API shape" here means, the imports of declaration files are opaque to importers, aren't they?

the concern is just you depend on a library X, it might be an implementation detail, that you may want to change without changing your API. the rational is, you can explicitly add the type annotation if you are fine having `X` as part of our API, or change it to something else.

**Comment by simonbuchan** [comment]
> the concern is just you depend on a library X, it might be an implementation detail, that you may want to change without changing your API. the rational is, you can explicitly add the type annotation if you are fine having X as part of our API, or change it to something else.

You mean when a JS client would be using undocumented internals, a TS client would get a fully typed public api for an internal dependency? I can see where someone might want that, but that would be something to solve when you have the ability for TS to know what an internal API even is! For example, if type declaration merging becomes a thing, then it makes sense that the merged file does not include declarations you didn't ask for. At the moment, without that, and the weird way package.json `types` field works, I don't think there's a way to have an NPM package with external submodules, which is a reasonable pattern - tree-shaking isn't permitted to remove module-init code, and packages can have different "modes" of use (`react-devtools-extension/logOnlyInProduction`, `core-js/lib/*`, `rxjs/operators`) 

So I would ask for the current 1:1 `.ts` to `.d.ts` model for `--declarations` to be "just make it work" in the absence of future options, given how awkward and arduous it can be to work with the current behavior.

Another option that keeps the current behavior without requiring the complexity of declaration merging may be: `--publicFiles`, which requires manual annotations only the exports of the listed files? And hopefully editors/doctools respect it and only expose those files (e.g. in path completion).

**Comment by unional** [comment]
I just found out that if I "compile" any nested interface can solve this issue.

I have this `tersify` package:
```ts
// index.ts
export interface TersifyOptions {
  maxLength: number
}

export interface Tersify {
  tersify(options?: Partial<TersifyOptions>): string
}
export type Tersible<T> = T & Tersify

export function tersible<T>(
  subject: T,
  tersify?: string | ((this: T, options: Partial<TersifyOptions>) => string))
  : Tersible<T> { ... }
```

When I consume this package:
```ts
import { tersible } from 'tersify'

export function create(...) { return tersible(...) }
```

I will encounter this issue.

But if I "compile" the `tersify` package:
```
export interface TersifyOptions {
  maxLength: number
}

export interface Tersify {
  tersify(options?: Partial<{ maxLength: number }>): string
}
export type Tersible<T> = T & {
  tersify(options?: Partial<{ maxLength: number }>): string
}

export function tersible<T>(
  subject: T, 
  tersify?: string | ((this: T, options: Partial<{ maxLength: number }>) => string))
  : Tersible<T> { ... }
```

Then the problem is gone.
It is gone because of the signature of `tersible()` references `Tersible<T>` which now does not reference any other types.

So the question is when the typing file is created, can we resolve all nested types so that there are no deep references?

At the end, it is a dynamic type system.
As long as the shape of the type matches, it should work fine.

Don't know if this makes sense, just want to throw it out here to open some discussion.

**Comment by simonbuchan** [comment]
@unional object types like `{ maxLength: number }` and interface types like `TersifyOptions` are subtly different in a few ways, so they can't be blindly exchanged in general: see #11207, #11961 at least, I've run into others too.

Even with manual work, this doesn't fix recursive types, since object types can't represent recursion:

```ts
export interface Tree<T> { value:T, left?: Tree<T>, right?: Tree<T> }
export function createTree<T>(values: Array<T>): Tree<T> { ... }
```

But a decent hack when it does work if you're willing to take the pain on the exporter side to save the importer!

**Comment by unional** [comment]
> But a decent hack when it does work if you're willing to take the pain on the exporter side to save the importer!

My stands with classical type system (interfaces) are that the whole thing is upside-down. It should be the consumer who defines the type instead of the author.

e.g.:
```ts
// authoring code: lightbulb
export interface LightBulb {
  on(): void
  off(): void
}

// consuming code: Lightswitch
export interface Switchable {
  on(): void
  off(): void
}
export class LightSwitch {
  constructor(private switchable: Switchable) { }
  flip() {
    this.up? this.switchable.on() : this.switchable.off();
  }
}
```

So to me, this is not really a hack, but how interfaces should really be. 🌷

**Comment by pauldraper** [comment]
So...
1. Adding import to declaration files will not always solve the problem. If a depends on b which depends on c, and implicitly exposes a type which came from node_modules/b/node_modules/c.d.ts, there is no good import which can be added (assuming 'b/node_modules/c' is not good ?).
1. But...many people (here and on #5711 and #9944) would like the TS compiler to take care of this anyway. In the problem case in `#1`, the declaration emitter could attempt to import a hypothetical c and complain if its does not exist or its type does not match. Automatically adding imports seems not unreasonable for the TS paradigm, given its capabilities for minimal type overhead and type inference. It would also reduce the overhead of using --noUnusedLocals.
1. Tooling would benefit. IDEs/linters struggle to identify "unused imports" because (a) it depends on whether declaration files will ever be emitted and (b) when declarations will be emitted, a full typechecked compile must happen before any import can properly be declared unused (and AFAIK none of them do this).
1. The compiler *can* do this, because (a) the declaration emitter already detects this and (b) the compiler has the info when compiling sans-declarations.
1. @weswigham is working on a new declaration emitter that doesn't have this feature, but if the feature were added, it would make sense to make sure it got added to the new emitter.

Questions:
* Is it agreed that `--autoDeclarationImports` would be a worthwhile addition?
* 
* What is the status on the new emitter?

**Comment by simonbuchan** [comment]
Not sure I follow 1? Here's the case I think you are describing.

```ts
// a.ts
import foo from "b";
// node_modules/b/index.ts
import C from "c";
export default class B extends C {};
// node_modules/b/node_modules/c/index.d.ts
export interface CArg { x: number }
export default declare class C<A = CArg> {}
```

TS 2.next could chose to emit `node_modules/b/index.d.ts` as:

```ts
import C, { CArg } from "c";
export default declare class B extends C {}
```

which current TS can understand fine when checking `a`, because it knows what path `B` was defined in, and thus where to find the module `c`. Am I missing something?

2. There's not really a "common cases" issue here - TS has [explicitly chosen](https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-348423398) not to support emitting types you didn't explicitly import for forward compat. reasons. It's just a bad intersection of concerns that any number of solutions could handle, they "just" need design and implementation.

4. I've not seen any problems with unused imports by IDE's? In the sense they report the same issues as `tsc --noUnusedLocals`.

I don't really like `--autoDeclarationImports` much, for the same reasons as TS team: I would prefer a more powerful solution that understands what modules are expected to be imported outside this compilation pass, and thus should have the manual export annotations; I just don't know what that looks like.

**Comment by pauldraper** [comment]
> Not sure I follow 1?+

Sorry, should have written this out.

```typescript
//a.ts
import b from "b";
export default function() {
    return b();
}

//node_modules/b/index.d.ts
import C from "c";
export default function(): C;

//node_modules/b/node_modules/c/index.d.ts
export default class {
}
```

I was unsure if what the import would be for a.d.ts. Would be it be?

```typescript
import C from 'b/node_modules/c';
export function(): C;
```

Is that import good? Not good?

> I've not seen any problems with unused imports by IDE's? In the sense they report the same issues as tsc --noUnusedLocals.

https://github.com/Microsoft/TypeScript/issues/5711#issuecomment-229497938
https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-265438354

If you use noUnusedLocals, you can be assured TS lint/Web Storm will not remove used imports for successfully compiling code. But that is now significantly more verbose https://github.com/Microsoft/TypeScript/issues/5711#issuecomment-287709774

---

On further thought, this is my proposal:

The compiler knows which file the needed symbol is in. If it is reachable from the source in the same module, in a direct dependency module, or via a path mapping, the compiler adds the import. If it can't, then have the same "cannot be named" error as today.

This covers common scenarios. And since it is always an improvement over the current situation, it doesn't even require a compiler flag. I can't see imagine a situation in which this would be harmful.

(Alternatively, we decide `import 'b/node_modules/c'` is always A-okay, and that becomes a possibility too.)

**Comment by simonbuchan** [comment]
Gotcha. Pretty sure it's not legal to do `b/node_modules/c` if `a` would be imported by someone else, because `npm` et. al. will flatten `a` `b` and `c` most of the time for a dependent of `a`, so `b/node_modules/c` would not find a legal module, e.g.:

```
// package.json
{ "dependencies": {"a": "*"} }
// node_modules/a/index.d.ts
import C from 'b/node_modules/c'; // Boom, module not found
// node_modules/b/index.d.ts
import C from 'c';
// node_modules/c/index.d.ts
export default ...
```

Importing from `c` directly is far more likely (but not guaranteed) to work. The safe thing to do is to explicitly re-export from `b`, and use that in `c`. Perhaps in a `declare module "c" { ... }`?

Yes, IDEs will warn on unused imports... just like `tsc --noUnusedLocals`. So I don't yet see it as an IDE issue (though they should update to handle whatever TS does here later)

The issue with requiring export typings is, to me, more that sometimes doing so can cause you to break when your dependencies update, which is ironic given the (valid) reasoning TS have for this being an issue in the first place!

**Comment by felixfbecker** [comment]
We just changed the return type of `graphql-tag` from `any` to `DocumentNode` from the `graphql` package, which because of this issue is a breaking change and needs to be postponed to the next major https://github.com/apollographql/graphql-tag/issues/150
It would be nice if operations like that were safe

**Comment by disintegrator** [comment]
I'm happy to import the types from external modules explicitly. I just wish there was a way to have `--noUnusedLocals` understand this scenario instead of flagging as error. This issue is quite important to library authors that want to use TypeScript to write library code that depends on third party modules.

**Comment by tvald** [comment]
> I'm happy to import the types from external modules explicitly. I just wish there was a way to have --noUnusedLocals understand this scenario instead of flagging as error.

It would be great if `--noUnusedLocals` would ignore names prefixed with an underscore, just like `--noUnusedParameters`. Then you could import type-only dependencies like so:

```
import { SomeType as _S } from 'other-package'
```

(This is currently possible with `tslint`'s equivalent [no-unused-variable](https://palantir.github.io/tslint/rules/no-unused-variable/) rule.)

**Comment by slashhuang** [comment]
same issue, hope somebody can make a merge request to get it done

**Comment by lostpebble** [comment]
This is becoming a major annoyance for me at the moment.

I've created an internal library for wrapping my functions for various purposes. This is basic example of something similar:

**Management.ts**
```
export interface IOutput {
  status: number;
  errorMessage?: string;
}

export type TFunction<I = {}, O extends IOutput = any> = (options: I) => Promise<O>;

export interface IFunctionDescription {
  title: string;
  description?: string;
  id?: string;
}

export function createFunction<I, O extends IOutput>(description: IFunctionDescription, func: TFunction<I, O>): TFunction<I, O> {
  return async (functionInput: I): Promise<O> => {
    console.log(`Running function ${description.title}`);

    const response = await func(functionInput);

    if (response.status === 200) {
      return response;
    }

    throw new Error(`Something went wrong: ${response.errorMessage}`);
  }
}
```
**MyColorFunction.ts**
```
import { createFunction, IOutput, TFunction } from "./Management";

export interface IOMyFunctionInput {
  color: string;
}

export interface IOMyFunctionOutput extends IOutput {
  isFavourite?: boolean;
}

const myColorFunction: TFunction<IOMyFunctionInput, IOMyFunctionOutput> = async ({ color }) => {
  if (color !== "blue") {
    return {
      status: 401,
      errorMessage: "Not a real color!",
    }
  }

  return {
    status: 200,
    isFavourite: true,
  }
}

export const myColorFunctionCreated = createFunction<IOMyFunctionInput, IOMyFunctionOutput>({
  title: "My Function For Favourite Colours",
}, myColorFunction)
```

Finally at the end I want to keep these functions neatly namespaced inside a kind of "Service" object:

**Usage.ts**
```
import { myColorFunctionCreated } from "./MyColorFunction";

export const FunctionService = {
  myColorFunction: myColorFunctionCreated,
}
```

And the resulting error:
```
Error:(3, 14) TS4023: Exported variable 'FunctionService' has or is using name 'IOMyFunctionInput' from external module "D:/.../MyColorFunction" but cannot be named.
Error:(3, 14) TS4023: Exported variable 'FunctionService' has or is using name 'IOMyFunctionOutput' from external module "D:/.../MyColorFunction" but cannot be named.
```

Now, I would expect Typescript to be able to infer those types because `createFunction<I, O>` is clear about what types it is using when exporting `myColorFunctionCreated`.

It doesn't even work if I be even more explicit on that export like this:

```
export const myColorFunctionCreated: TFunction<IOMyFunctionInput, IOMyFunctionOutput> = createFunction<IOMyFunctionInput, IOMyFunctionOutput>({
  title: "My Function For Favourite Colours",
}, myColorFunction)
```

The only way to mitigate this (and it gets very tedious very quickly - not to mention makes my code very unreadable after a while) is this:

**Usage.ts**
```
import { IOMyFunctionInput, IOMyFunctionOutput, myColorFunctionCreated } from "./MyColorFunction";
import { TFunction } from "./Management";

export const FunctionService = {
  myColorFunction: myColorFunctionCreated as TFunction<IOMyFunctionInput, IOMyFunctionOutput>,
}
```

Or, of course, to just have those imports in the file and not explicitly make use of them. But that causes other issues like others have mentioned (unused imports) and in most IDEs will complain or refactor out automatically because they are not being used.

Basically, we are required to redefine types at the latest stage of usage - even though these types have a very clear pathway to exactly what they are. This is really dirtying up the code base with unnecessary declarations all over the place.

**Comment by pmoleri** [comment]
Hi there,

I'm in the same boat here. To work around this issue, I had to add imports in several files and exports to prevent unused warning of those imports.

But, in my case, my package exports only a small surface where most of these types don't even appear. So, I had polluted my package's internal code, when I only need declarations for the entry point.
Is there any way to instruct typescript to generate type declarations only for the entry point and its exports?

**Comment by weswigham** [comment]
@pmoleri yes, there is - disable declarations on your main build, then add a second build with `declarationsOnly` set, and only include your entry point file, rather than all of your TS in that build.

**Comment by pmoleri** [comment]
Thanks @weswigham , that's a good idea. I also found another workaround, using `--stripInternal` experimental flag.

This is my attempt to summarize of workarounds:

 - Import the missing types. This probably yields unused errors, so you may add an export:
```
import { type } from "./other-module";
export { type }
```

 - If your module is internal (i.e. it's not exported directly or indirectly from the package entry point), there are 2 additional alternatives:
       a. Emit declarations in a separate step including only public files ([previous comment](#issuecomment-383995178)).
      b. Use experimental compiler flag `--stripInternal`  and add comments to your exports:
      ```
      /** @internal TS issue 9944 */
      export class Project {
          ...
      ```

**Comment by kohlmannj** [comment]
As a relative newcomer to TypeScript, it's dismaying to see this two-year-old issue still in the discussion phase.

A slice of my experience: TSLint's [`no-unused-variable`](https://palantir.github.io/tslint/rules/no-unused-variable/) rule is reliant on the TypeScript compiler's `noUnusedLocals` option ([example scenario with `React`](https://github.com/palantir/tslint/issues/589)), but enabling said option generates false positives for imports of external type definitions.

This feels like a kind of design stalemate, which leaves me without any clear sense of which code quality tool\* or configuration settings might allow me to reliably assess unused imports and local variables in [my own TypeScript source files](https://github.com/kohlmannj/limina) while also maintaining comprehensive type definitions.

Going all the way back to @gund's [comment](https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-289669372) and @disintegrator's [comment](https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-367688820), it seems like one of two things need to happen:

- **Option 1** (my preference): using `"declaration": true` and `"noUnusedLocals": true` prevents false positives for external type definition imports required to name any in-scope types
- **Option 2**: a third flag (e.g. @gund's `"implicitDeclaration": true`) be added to explicitly cover this scenario

—

\* At some point I had ESLint + TypeScript working, but I'm not convinced that ESLint + Babel 7 + TypeScript + [eslint-typescript-parser](https://github.com/eslint/typescript-eslint-parser) is a reliable developer experience as of this writing (May 2018).

**Comment by asvetliakov** [comment]
Still doesn't work for me:
Typescript: 3.0.0-dev.20180522

Repro:
file1.ts
```ts
import Color from "color";
export declare function styled(): Color;
```

file2.ts
```ts
import { styled }  from "./styled";
export const A = styled();
```

Exported variable 'A' has or is using name 'Color' from external module "/Users/asvetl/work/test/node_modules/@types/color/index" but cannot be named.

**Comment by weswigham** [comment]
What's the type of the `Color` thing being imported?

**Comment by asvetliakov** [comment]
@weswigham It's ![DT color](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/color/index.d.ts) package (using allowSyntheticDefaultImports: true ).

This also doesn't work:

color.ts
```ts
interface Color {
    c: string;
}
export default Color;
```

file1.ts
```
import Color from "./color";
```

but this works:
```ts
export default interface Color {
     c: string;
}
```

**Comment by mhegazy** [comment]
Mind filing a different ticket for this one.

**Comment by kohlmannj** [comment]
Since this issue has been closed, here's some follow-up from my previous comment for those who read this thread: https://github.com/Microsoft/TypeScript/issues/9944#issuecomment-385599123

One of my concerns, which led me to this issue, was that TypeScript's "cannot be named" error was telling me to import additional types. When I did so, this led to warnings about those imported types not being used, since I have `"noUnusedLocals": true` set in my tsconfig.json.

_The solution_ was to explicitly declare my types. In VSCode, you can even hover over a symbol, **select the text** in of the tooltip showing you that symbol's automatically inferred type, and paste it into your source file.

_Et voilà_ — at this point, you've referenced the imported type locally, so the "unused local" warning disappears.

**Comment by mprobst** [comment]
@weswigham I think this is fixed with TS2.9, at least as far as our use cases are concerned.

**Comment by trusktr** [comment]
@weswigham I'm on typescript `v3.3.0-dev.20181129`, and it looks like this issue is not completely fixed.

The following issue has a sample of what my code looks like along with the problem described here: https://github.com/Microsoft/TypeScript/issues/28754#issuecomment-443820268

Unfortunately I am not able to make a small reproduction, and I can't share the larger private code where it is happening.

The strange thing is that it only happens in _some_ files, but not all of them.

**Comment by hutber** [comment]
For me, the error was simply that I had no declared an import at all.

`import styled from 'styled-components'`
`styled` was completely undefined, but instead of a normal JS error I got a TS error.

**Comment by 8bu** [comment]
Don't wanna bring this issue up again but I'm really struggling with this error. I read through the comments but can't really understand my problem & solution provided above.

### The error
```
Exported variable 'useWeb3' has or is using name 'RedBN' from external module "~/sample-project/node_modules/@types/bn.js/index" but cannot be named.
```

### My code:

```
example.ts
import Web3 from 'web3';
import { ref, toRefs } from 'vue';

const web3 = ref<null|Web3>(null);
export const useWeb3 = () => {
  return {
    state: toRefs(web3)
  }
}
```

From what I've read in this issue, I just need to import the `RedBN` type to my `example.ts` ? But RedBN was declare locally & have no export then how can I resolve this problem..

**Comment by jcbdev** [comment]
@8bu, @kohlmannj, @trusktr 

I had this for an exported type which is actually just a renamed internal interface which could not be imported/re-exported so I found a work around:

```
// TSESTreeOptions is exported from library but its actually just a rename of an internal type ParseAndGenerateServicesOptions which is not exported and this causes the error in type resolution
import { TSESTreeOptions as _TSESTreeOptions } from '@typescript-eslint/typescript-estree';

//Clone type and use this type in place of the original for my modules export api/signature
export type TSESTreeOptions = { [K in keyof _TSESTreeOptions]: _TSESTreeOptions[K] }; 

export function doSomething(options: TSESTreeOptions) {
    callFunctionThatRequiresOriginalType(options); //Type resolution should know types are compatible
    callFunctionThatRequiresOriginalType(options as _TSESTreeOptions); //If not, cast to original type
}
```

This is the original type export from the `@typescript-eslint/typescript-estree` library. 
```
export declare type TSESTreeOptions = ParseAndGenerateServicesOptions;
```

but if you find `ParseAndGenerateServicesOptions` its an internal interface
```
interface ParseAndGenerateServicesOptions extends ParseOptions {
```

My theory is this type of export is the issue as typescript sees that TSESTreeOptions = ParseAndGenerateServicesOptions and potentially just treats it as an alias, or it doesn't work when "exported type = internal interface"?  Then when it tries to resolve this alias if you import and then export from another module further down the line the alias cannot be resolved because the type is resolves is internal?

**Comment by btakita** [comment]
This issue makes opaque types very difficult to use in some situations.

```ts
declare const norm1_5_sym:unique symbol
export type norm1_5_T<Base extends number = number> =
	Base extends score_T
	? score1_5_T
	: Base extends percentile_T
		? percentile1_5_T
		: Base&{ [norm1_5_sym]:any }
```
```ts
// TS4023: Exported variable 'norm1_5_M_color' has or is using name 'norm1_5_sym' from external module "..." but cannot be named.
export const norm1_5_M_color = Object.freeze(
	new Map<norm1_5_T, string>([
		[null as norm1_5_T, norm1_5__null__color],
		[1 as norm1_5_T, norm1_5__1__color],
		[2 as norm1_5_T, norm1_5__2__color],
		[3 as norm1_5_T, norm1_5__3__color],
		[4 as norm1_5_T, norm1_5__4__color],
		[5 as norm1_5_T, norm1_5__5__color],
	]))
```

I have not been able to export/import `norm1_5_sym` to make the error go away.

**Comment by icopp** [comment]
I don't understand why this is closed. It's still an issue, and the different behavior here between `type` and `interface` is obviously a bug from the end-user perspective.

**Comment by cwadrupldijjit** [comment]
I just encountered this today.  I wasn't using a very old version of TypeScript (_much_ later than this issue), so I was surprised to see this error suddenly crop up in a file I wrote not long ago.

Considering the age of this issue, I think I'll have to create a new one to figure out what I'm supposed to do and perhaps a bit of conversation on why I prefer the approach I took rather than the suggested approach.  And perhaps there's already another conversation about it.

**Comment by RyanCavanaugh** [comment]
Locking due to drive-by comments.

Please read https://github.com/Microsoft/TypeScript/wiki/FAQ#the-inferred-type-of-x-cannot-be-named-without-a-reference-to-y-this-is-likely-not-portable-a-type-annotation-is-necessary and log a new issue if you encounter behavior not covered by that FAQ entry.
